OSD: Optimised EC Deep Scrubbing Implementation

JonBailey1993 opened the pull request:
depends on #62952

Initial implementation of deep-scrubbing for Optimised EC. Performs deep scrubbing by taking the digests and putting them through the erasure coding plugin to decide if the calculated parity digest is the same as the actual parities digest. Added a unit test to TestErasureCodePlugins.cc to verify this functionality, and added unit tests to test_scrubber_be.cc to test the new scrubbing behaviour.

I have still got some small efficiency and tidying up, but functionally it should be sound and wanted to be able for people to take a look.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

aainscow commented:
Did you mean to add us as reviewers, rather than assign to us?

aainscow reviewed (approved):
Mostly nits.  Some concerns:

1. There are places where we are using <k to mean "is data" this will break for LRC... academic. 
2. We seem to be supporting legacy EC for some stuff. This adds test burden and we only want if we have time for that. 
3. ISA has a m==1 restriction - given this is our default plugin, this is very worrying.

aainscow reviewed (changes requested):
Meant to click "request changes"

JonBailey1993 commented on src/erasure-code/isa/ErasureCodeIsa.h:
```diff
@@ -73,6 +73,10 @@ class ErasureCodeIsa : public ceph::ErasureCode {
         technique == "default"sv) {
       flags |= FLAG_EC_PLUGIN_OPTIMIZED_SUPPORTED;
     }
+    else if (technique != "cauchy"sv && m == 1)
```
The condition is wrong, I think the if statement above was meant to also contain the flag setting as the reed_sol_van works for any value of m. Will update and also make inclusive as you suggest.

aainscow replied:
Phew!

JonBailey1993 commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -92,17 +93,32 @@ ScrubBackend::ScrubBackend(ScrubBeListener& scrubber,
+  // Lengths of zero buffers need to match the length of the shard.
+  // So we initialise a new buffer to the correct length per shard.
+  bufferlist zero_bl;
+  zero_bl.append(buffer::create(
```
That interface sounds better, will update.

JonBailey1993 commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -502,6 +547,74 @@ auth_selection_t ScrubBackend::select_auth_object(const hobject_t& ho,
   }
 
+  if (auth_version != eversion_t() && m_is_optimized_ec &&
+      m_pg.get_ec_supports_crc_encode_decode() && available_shards.size() != 0) {
```
Will update

JonBailey1993 replied:
Have added a check for !m_is_replicated in cases where this wasn't checked previously, as I specifically made m_pg.get_ec_supports_crc_encode_decode() not support replicated pools

JonBailey1993 commented:
> Mostly nits. Some concerns:
> 
> 1. There are places where we are using <k to mean "is data" this will break for LRC... academic.
> 2. We seem to be supporting legacy EC for some stuff. This adds test burden and we only want if we have time for that.
> 3. ISA has a m==1 restriction - given this is our default plugin, this is very worrying.

Thanks for raising these points.
1: Thanks for pointing out this, I will go through and update these to use better interfaces that are "LRC compatible".
2: I think from a conceptual standpoint everything we do for optimised EC should work on legacy EC in terms of these deep-scrubbing changes. However the point of view of the test burden is not something I had considered. I will make a branch in git of the current implementation with the legacy so that if we want to reintroduce these again we will not lose the code, however for the time being I think I will remove the Legacy EC changes from this PR and just have it concentrate on doing this work for Optimised EC.
3: This was a mistake on my part. The m == 1 restriction should apply to Cauchy ONLY. Reed Solomon Van should work with any m value. I will update to reflect this. I will get this fixed.

JonBailey1993 commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -1057,7 +1174,28 @@ ScrubBackend::auth_and_obj_errs_t ScrubBackend::match_in_shards(
+                                                     srd);
+
+      if (!m_is_replicated && m_is_optimized_ec &&
+          m_pg.get_ec_supports_crc_encode_decode()) {
```
Fixed

JonBailey1993 commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -1143,6 +1281,78 @@ ScrubBackend::auth_and_obj_errs_t ScrubBackend::match_in_shards(
   }
 
+  if (!m_is_replicated && m_is_optimized_ec &&
+      m_pg.get_ec_supports_crc_encode_decode())
```
Fixed

JonBailey1993 commented on src/osd/ECBackend.cc:
```diff
@@ -1822,56 +1870,59 @@ int ECBackend::be_deep_scrub(
   ECUtil::HashInfoRef hinfo = unstable_hashinfo_registry.get_hash_info(
     poid, false, o.attrs, o.size);
+  if (sinfo.supports_ec_optimisations() && sinfo.supports_encode_decode_crcs()) {
+    // HInfo most likely does not exist when ec optimisations are enabled
```
Fixed

JonBailey1993 commented on src/osd/ECBackendL.cc:
```diff
@@ -1685,6 +1685,45 @@ void ECBackendL::objects_read_async(
 }
 
+  bool ECBackendL::ec_can_decode(const std::set<int> &available_shards) const {
+  std::map<int, std::vector<std::pair<int, int>>> minimum_set;
```
Answered here: https://github.com/ceph/ceph/pull/63184#issuecomment-2866608121

JonBailey1993 commented on src/osd/ECSwitch.h:
```diff
@@ -328,6 +335,65 @@ class ECSwitch : public PGBackend
   }
 
+  bool get_ec_supports_crc_encode_decode() const override
+  {
```
Answered here: https://github.com/ceph/ceph/pull/63184#issuecomment-2866608121

JonBailey1993 commented on src/test/osd/scrubber_generators.h:
```diff
@@ -203,6 +211,13 @@ static inline RealObj crpt_do_nothing(const RealObj& s, int osdn)
 }
 
+  static inline RealObj crpt_object_hash(const RealObj& s, [[maybe_unused]] int osdn)
+{
```
Fixed

JonBailey1993 commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -502,6 +547,74 @@ auth_selection_t ScrubBackend::select_auth_object(const hobject_t& ho,
+      // number of parity shards
+
+      int missing_shards = 0;
+      for (shard_id_t shard_id; shard_id < m_pg.get_ec_data_chunk_count();
```
This is now updated to use this set

JonBailey1993 commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -502,6 +547,74 @@ auth_selection_t ScrubBackend::select_auth_object(const hobject_t& ho,
+      }
+
+      bufferlist crc_bl;
+      for (shard_id_t shard_id; shard_id < m_pg.get_ec_data_chunk_count();
```
Done

JonBailey1993 commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -1143,6 +1281,78 @@ ScrubBackend::auth_and_obj_errs_t ScrubBackend::match_in_shards(
+      // a data consistency issue that should be reported.
+      for (auto& [srd, bl] : digests)
+      {
+        if (srd < m_pg.get_ec_data_chunk_count())
```
Done

JonBailey1993 commented on src/osd/ECBackend.cc:
```diff
@@ -1722,6 +1722,61 @@ void ECBackend::objects_read_async(
+  }
+
+  if (ec_impl->get_profile().find("profile")->second == "clay") {
+    // FIXME: Subchunks can be passed through to the minimum_to_decode function
```
Now uses `sinfo.supports_sub_chunks()`

JonBailey1993 commented on src/test/osd/scrubber_generators.cc:
```diff
@@ -40,16 +40,33 @@ std::pair<bufferlist, std::vector<snapid_t>> create_object_snapset(
 
   for (auto osd : active_osds) {
+    shard_id_t shard = std::find_if(acting_shards.begin(),acting_shards.end(), [&osd](pg_shard_t pg_shard)
+    {
```
Have adjusted, should be fixed?

JonBailey1993 replied:
Have run git-clang-format to fix formatting and whitespace.

JonBailey1993 commented on src/osd/ECBackendL.cc:
```diff
@@ -1685,6 +1685,45 @@ void ECBackendL::objects_read_async(
+}
+
+  std::map<int, bufferlist> ECBackendL::ec_encode_acting_set(
+      const bufferlist &in_bl) const {
```
Function removed. No longer relevant.

JonBailey1993 commented on src/osd/ECBackendL.cc:
```diff
@@ -1685,6 +1685,45 @@ void ECBackendL::objects_read_async(
+}
+
+  std::map<int, bufferlist> ECBackendL::ec_decode_acting_set(
+      const std::map<int, bufferlist> &shard_map, int chunk_size) const {
```
Function removed. No longer relevant.

JonBailey1993 commented on src/test/osd/scrubber_test_datasets.cc:
```diff
@@ -55,10 +62,19 @@ SnapsetMockData::CookedCloneSnaps ms1_fn()
+SnapsetMockData::CookedCloneSnaps ms2_fn()
+{
+	return {{}, {}, {}};
+}
```
Fixed

JonBailey1993 commented on src/osd/ECBackend.h:
```diff
@@ -437,15 +442,31 @@ class ECBackend : public ECCommon {
   }
 
+  unsigned get_ec_stripe_width() const {
+    return sinfo.get_k() + sinfo.get_m();
```
This function no longer exists, have changed to using k_plus_m() of sinfo now.

JonBailey1993 commented on src/test/osd/scrubber_test_datasets.cc:
```diff
@@ -116,4 +132,14 @@ ScrubGenerator::RealObjsConf minimal_snaps_configuration{
 };
 
+ScrubGenerator::RealObj erasure_code_obj{ghobject_t{ec_hobj_ms1, 1, shard_id_t{0}},
+				 RealData{100,
```
Have run git-clang-format to fix formatting and whitespace.

ronen-fr commented on src/erasure-code/isa/ErasureCodeIsa.h:
```diff
@@ -69,9 +69,11 @@ class ErasureCodeIsa : public ceph::ErasureCode {
-    if (technique == "reed_sol_van"sv ||
-        technique == "default"sv) {
-      flags |= FLAG_EC_PLUGIN_OPTIMIZED_SUPPORTED;
+    if (technique == "reed_sol_van"sv) {
```
(a side note: we should consider changing 'technique' to std::string)

ronen-fr replied:
(just a note: I wanted to suggest to compare, now that 'technique' is a string, to ".."s and not ".."sv.
But for some reason - the binary produced is tighter on the 'sv'. Interesting.)

ronen-fr commented on src/erasure-code/isa/ErasureCodePluginIsa.cc:
```diff
@@ -42,11 +42,13 @@ int ErasureCodePluginIsa::factory(const std::string &directory,
+                                            ErasureCodeIsaDefault::kVandermonde,
+                                            t.c_str());
+} else {
+if ((t == "cauchy")) {
```
nit: indentations

JonBailey1993 replied:
Fixed. Also changed the above else into an else if as I saw no reason why it shouldn't have been that before. Looking at the flow of the code, it seemed to make more sense as else if and reducing the indentation by one.

ronen-fr commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -419,10 +434,35 @@ auth_selection_t ScrubBackend::select_auth_object(const hobject_t& ho,
   /// This creates an issue with 'digest_match' that should be handled.
   std::list<pg_shard_t> shards;
+  shard_id_set available_shards;
+  unsigned int digestMapSize = 0;
```
id format

ronen-fr replied:
(and - if possible - specify the type explicitly, i.e. uint32_t or 64)

JonBailey1993 replied:
When you say id format, are you talking about using uint32_t like you say in the second comment, or are you talking about something to do with the "shard_id_set" format?

ronen-fr replied:
trying to say that 'camelcase is ugly' :-)

JonBailey1993 replied:
Ohh, understood. Have updated - I missed that one on my pass through :-)

ronen-fr commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -419,10 +434,35 @@ auth_selection_t ScrubBackend::select_auth_object(const hobject_t& ho,
+          m_pg.ec_get_sinfo().get_chunk_size());
+      b.copy_in(0, length, crc_bytes);
+      bufferlist crc_list;
+      crc_list.append(b);
```
would it be better to append directly into digest_map[s] (after having it created)?
i.e. something like:
digest_map[srd.shard] = bufferlist{};
digest_map[srd.shard].append(b);
?

JonBailey1993 replied:
Have updated.

ronen-fr commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -509,6 +549,71 @@ auth_selection_t ScrubBackend::select_auth_object(const hobject_t& ho,
+        uint32_t zero_data_crc = generate_zero_buffer_crc(
+            shard_id, logical_to_ondisk_size(ret_auth.auth_oi.size, shard_id));
+        for (long unsigned int i = 0; i < sizeof(zero_data_crc); i++) {
+          digest_map[shard_id].c_str()[i] =
```
@rzarzynski - any problem here (after your buffer changes)? thanks

ronen-fr commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -509,6 +549,71 @@ auth_selection_t ScrubBackend::select_auth_object(const hobject_t& ho,
+        }
+      }
+
+      int num_redundancy_shards = (m_pg.ec_get_sinfo().get_m());
```
maybe 'const int' ? and why the enclosing ()?

JonBailey1993 replied:
The enclosing () were because it was a calculation during development which I preferred having the brackets for - and I modified to a single getter, but missed removing the enclosing brackets when I did so. I have removed them now and made this into a const.

ronen-fr commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -509,6 +549,71 @@ auth_selection_t ScrubBackend::select_auth_object(const hobject_t& ho,
+      // number of parity shards
+
+      int missing_shards = 0;
+      for (shard_id_t shard_id : m_pg.ec_get_sinfo().get_data_shards()) {
```
using an existing algorithm?

JonBailey1993 replied:
Changed to use std::count_if.

JonBailey1993 commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -509,6 +549,71 @@ auth_selection_t ScrubBackend::select_auth_object(const hobject_t& ho,
+      }
+
+      bufferlist crc_bl;
+      for (shard_id_t shard_id : m_pg.ec_get_sinfo().get_data_shards()) {
```
Have updated

JonBailey1993 commented:
make check (arm64)
`Reading package lists...
E: Write error - write (28: No space left on device)
E: IO Error saving source cache
E: The package lists or status file could not be parsed or opened.
dpkg: unrecoverable fatal error, aborting:
 unable to fill /var/lib/dpkg/updates/tmp.i with padding: No space left on device
CI_DEBUG: trying to install ccache git debianutils clang lvm2 again `

Looks like disk ran out of space during what looks like CMake. Will retrigger.

JonBailey1993 commented:
jenkins test make check arm64

ronen-fr commented on src/erasure-code/isa/ErasureCodeIsa.h:
```diff
@@ -53,10 +53,10 @@ class ErasureCodeIsa : public ceph::ErasureCode {
   uint64_t flags;
 
-  ErasureCodeIsa(const char *_technique,
+  ErasureCodeIsa(const std::string &_technique,
```
+1

ronen-fr commented on src/erasure-code/isa/ErasureCodeIsa.h:
```diff
@@ -151,15 +153,14 @@ class ErasureCodeIsaDefault : public ErasureCodeIsa {
-  ErasureCodeIsa("default", _tcache),
-  encode_coeff(0), encode_tbls(0)
-  {
+        ErasureCodeIsa(technique, _tcache),
```
indentation problem?

JonBailey1993 replied:
Should now be fixed

ronen-fr commented:
LGTM. Would like to have Sam's review, too.

ronen-fr reviewed (approved):
I'd appreciate an additional review by @athanatos

JonBailey1993 commented:
Rebased to include latest changes to #62952
(also added comments describing the new functions I have added in scrubber_common.cc (PgScrubBeListener) at the same time. No code changes made.)

athanatos reviewed (approved):
LGTM

JonBailey1993 commented:
Make check failure:

`  Scenario: "Test IOPS Read"                           # tests_dashboards/features/ceph-cluster.feature:225
    Given the following series                         # tests_dashboards/features/environment.py:66
      | metrics                                                    | values    |
      | ceph_osd_op_r{job="ceph",cluster="mycluster", osd="osd.0"} | 1000 1000 |
      | ceph_osd_op_r{job="ceph",cluster="mycluster", osd="osd.1"} | 2000 1500 |
    Then Grafana panel `IOPS` with legend `Read` shows # tests_dashboards/features/environment.py:106
      | metrics | values |
      | {}      | 25     |
Unit Testing:  /tmp/tmp56ji56wg
  FAILED:
    expr: "Sum(rate(ceph_pool_rd{cluster=~\"mycluster\",}[1m]) *\n  on(pool_id) group_left(instance,name) ceph_pool_metadata{cluster=~\"mycluster\",name=~\"UNSET VARIABLE\"})", time: 1m,
        exp:"{} 2.5E+01"
        got:"nil"`

Failure in tests_dashboards which has not been modified in this PR, looks like test is unhappy about unset variable in response to query. Possible flaky unit test, will retrigger.

Make check (arm64) failure:

`The package has been created.
Attention, the package has been created in the current directory,
not in ".." as indicated by the message above!
dpkg: unrecoverable fatal error, aborting:
 unable to fill /var/lib/dpkg/updates/tmp.i with padding: No space left on device`

Failure during build process due to device being out of space. Will retrigger.

JonBailey1993 commented:
jenkins test make check

JonBailey1993 commented:
jenkins test make check arm64

JonBailey1993 commented:
The above changes I just pushed are just updating the wording of the error strings given by the scrubbing. No other changes were made.

JonBailey1993 commented:
Rebased on main

JonBailey1993 commented:
make check:

99% tests passed, 1 tests failed out of 317

Total Test time (real) = 7250.30 sec

The following tests FAILED:
	 85 - unittest_ceph_assert (Timeout)
Errors while running CTest

Unit test timed out. Will retrigger.

make check (arm64):

`npm ERR! code ECONNRESET
npm ERR! network aborted
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network 
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'`

Looks like network issue occurred during build. Will retrigger.

JonBailey1993 commented:
jenkins test make check

JonBailey1993 commented:
jenkins test make check arm64

JonBailey1993 commented:
jenkins test make check

ljflores commented:
@JonBailey1993 Apologies for the delay in QA- I ran into some issues that I think might be related to this PR. I'm asking Bharath to drop it and rerun some jobs to make sure. Can you take a look at the following issues and let me know what you think? I particularly think that (1) is related since the job failed twice, and it is running an ec-radosbench workload. (2) might not be related since it passed in a rerun, and I've seen a similar issue before, but it would be good to make sure.

Testing ref: https://tracker.ceph.com/issues/71409

### Suspected regressions:

1. This one failed twice:

/a/skanta-2025-06-03_11:03:28-rados-wip-bharath1-testing-2025-06-02-2052-distro-default-smithi/8307501
/a/skanta-2025-06-05_23:47:06-rados-wip-bharath1-testing-2025-06-02-2052-distro-default-smithi/8311804

rados/thrash-erasure-code/{ceph clusters/{fixed-4} fast/fast mon_election/connectivity msgr-failures/fastclose objectstore/{bluestore/{alloc$/{avl} base mem$/{low} onode-segment$/{512K-onoff} write$/{v1/{compr$/{no$/{no}} v1}}}} rados recovery-overrides/{more-async-partial-recovery} supported-random-distro$/{ubuntu_latest} thrashers/default thrashosds-health workloads/ec-radosbench}

```
2025-06-03T17:41:21.543 DEBUG:teuthology.orchestra.run.smithi119:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 30 ceph --cluster ceph --admin-daemon /var/run/ceph/ceph-osd.13.asok dump_ops_in_flight
2025-06-03T17:41:21.543 INFO:tasks.ceph.osd.10.smithi144.stderr:2025-06-03T17:41:21.534+0000 7f6eff0b5640 -1 received  signal: Hangup from /usr/bin/python3 /usr/bin/daemon-helper kill ceph-osd -f --cluster ceph -i 10  (PID: 26881) UID: 0
2025-06-03T17:41:21.581 INFO:teuthology.orchestra.run.smithi044.stderr:nodeep-scrub is set
2025-06-03T17:41:21.643 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_teuthology_eaeb97003cfc43fc86754e4e45e7b398c784dedf/teuthology/run_tasks.py", line 105, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_teuthology_eaeb97003cfc43fc86754e4e45e7b398c784dedf/teuthology/run_tasks.py", line 83, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_teuthology_eaeb97003cfc43fc86754e4e45e7b398c784dedf/teuthology/task/full_sequential.py", line 37, in task
    mgr.__exit__(*exc_info)
  File "/usr/lib/python3.10/contextlib.py", line 142, in __exit__
    next(self.gen)
  File "/home/teuthworker/src/github.com_ceph_ceph-c_28e985f5708ff44ef7be57b0a26d7b94fad5c783/qa/tasks/radosbench.py", line 142, in task
    run.wait(radosbench.values(), timeout=timeout)
  File "/home/teuthworker/src/git.ceph.com_teuthology_eaeb97003cfc43fc86754e4e45e7b398c784dedf/teuthology/orchestra/run.py", line 479, in wait
    check_time()
  File "/home/teuthworker/src/git.ceph.com_teuthology_eaeb97003cfc43fc86754e4e45e7b398c784dedf/teuthology/contextutil.py", line 134, in __call__
    raise MaxWhileTries(error_msg)
teuthology.exceptions.MaxWhileTries: reached maximum tries (800) after waiting for 4800 seconds
```

2. This one passed on a rerun:

/a/skanta-2025-06-03_11:03:28-rados-wip-bharath1-testing-2025-06-02-2052-distro-default-smithi/8307584
```
rados/thrash/{0-size-min-size-overrides/2-size-2-min-size 1-pg-log-overrides/normal_pg_log 2-recovery-overrides/{more-async-partial-recovery} 3-scrub-overrides/{max-simultaneous-scrubs-1} backoff/normal ceph clusters/{fixed-4} crc-failures/bad_map_crc_failure d-balancer/upmap-read mon_election/classic msgr-failures/osd-delay msgr/async-v1only objectstore/{bluestore/{alloc$/{stupid} base mem$/{low} onode-segment$/{256K} write$/{v2/{compr$/{no$/{no}} v2}}}} rados supported-random-distro$/{centos_latest} thrashers/pggrow_host thrashosds-health workloads/radosbench}

2025-06-03T19:45:14.007 INFO:tasks.ceph.ceph_manager.ceph:{'pgid': '6.a', 'version': "0'0", 'reported_seq': 1228, 'reported_epoch': 2255, 'state': 'undersized+peered', 'last_fresh': '2025-06-03T19:45:07.345734+0000', 'last_change': '2025-06-03T19:25:07.912242+0000', 'last_active': '2025-06-03T19:25:06.897205+0000', 'last_peered': '2025-06-03T19:45:07.345734+0000', 'last_clean': '2025-06-03T19:25:06.897205+0000', 'last_became_active': '0.000000', 'last_became_peered': '2025-06-03T19:25:07.912242+0000', 'last_unstale': '2025-06-03T19:45:07.345734+0000', 'last_undegraded': '2025-06-03T19:45:07.345734+0000', 'last_fullsized': '2025-06-03T19:25:07.911604+0000', 'mapping_epoch': 1632, 'log_start': "0'0", 'ondisk_log_start': "0'0", 'created': 1632, 'last_epoch_clean': 0, 'parent': '0.0', 'parent_split_bits': 0, 'last_scrub': "0'0", 'last_scrub_stamp': '2025-06-03T19:25:06.897205+0000', 'last_deep_scrub': "0'0", 'last_deep_scrub_stamp': '2025-06-03T19:25:06.897205+0000', 'last_clean_scrub_stamp': '2025-06-03T19:25:06.897205+0000', 'objects_scrubbed': 0, 'log_size': 0, 'log_dups_size': 0, 'ondisk_log_size': 0, 'stats_invalid': False, 'dirty_stats_invalid': False, 'omap_stats_invalid': False, 'hitset_stats_invalid': False, 'hitset_bytes_stats_invalid': False, 'pin_stats_invalid': False, 'manifest_stats_invalid': False, 'snaptrimq_len': 0, 'last_scrub_duration': 0, 'scrub_schedule': 'queued for scrub', 'scrub_duration': 0, 'objects_trimmed': 0, 'snaptrim_duration': 0, 'stat_sum': {'num_bytes': 0, 'num_objects': 0, 'num_object_clones': 0, 'num_object_copies': 0, 'num_objects_missing_on_primary': 0, 'num_objects_missing': 0, 'num_objects_degraded': 0, 'num_objects_misplaced': 0, 'num_objects_unfound': 0, 'num_objects_dirty': 0, 'num_whiteouts': 0, 'num_read': 0, 'num_read_kb': 0, 'num_write': 0, 'num_write_kb': 0, 'num_scrub_errors': 0, 'num_shallow_scrub_errors': 0, 'num_deep_scrub_errors': 0, 'num_objects_recovered': 0, 'num_bytes_recovered': 0, 'num_keys_recovered': 0, 'num_o
bjects_omap': 0, 'num_objects_hit_set_archive': 0, 'num_bytes_hit_set_archive': 0, 'num_flush': 0, 'num_flush_kb': 0, 'num_evict': 0, 'num_evict_kb': 0, 'num_promote': 0, 'num_flush_mode_high': 0, 'num_flush_mode_low': 0, 'num_evict_mode_some': 0, 'num_evict_mode_full': 0, 'num_objects_pinned': 0, 'num_legacy_snapsets': 0, 'num_large_omap_objects': 0, 'num_objects_manifest': 0, 'num_omap_bytes': 0, 'num_omap_keys': 0, 'num_objects_repaired': 0}, 'up': [7], 'acting': [7], 'avail_no_missing': ['7'], 'object_location_counts': [], 'blocked_by': [], 'up_primary': 7, 'acting_primary': 7, 'purged_snaps': []}
2025-06-03T19:45:14.010 INFO:tasks.thrashosds.thrasher:Traceback (most recent call last):
  File "/home/teuthworker/src/github.com_ceph_ceph-c_28e985f5708ff44ef7be57b0a26d7b94fad5c783/qa/tasks/ceph_manager.py", line 192, in wrapper
    return func(self)
  File "/home/teuthworker/src/github.com_ceph_ceph-c_28e985f5708ff44ef7be57b0a26d7b94fad5c783/qa/tasks/ceph_manager.py", line 1488, in _do_thrash
    self.ceph_manager.wait_for_recovery(
  File "/home/teuthworker/src/github.com_ceph_ceph-c_28e985f5708ff44ef7be57b0a26d7b94fad5c783/qa/tasks/ceph_manager.py", line 3015, in wait_for_recovery
    assert now - start < timeout, \
AssertionError: wait_for_recovery: failed before timeout expired
```

ronen-fr commented:
RADOS-approved (see https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues71975).
But note that a rebase is needed.

JonBailey1993 commented:
Rebase above fixes marge conflicts from all_shards being added into ECUtil.h while this PR was in progress (This PR had done the same and it conflicted). Have removed my additions and used the existing implementations of all_shards instead.
